packages/roosterjs-editor-plugins/lib/PasteManager/PasteManager.ts

```diff
@@ -59,129 +73,77 @@ export default class PasteManager implements EditorPlugin {
-                processImages(container, clipboardData);
-                clipboardData.htmlData = normalizeContent(container.innerHTML);
-
-                let document = this.editor.getDocument();
-                let fragment = document.createDocumentFragment();
-                let nodes = fromHtml(clipboardData.htmlData, document);
-                for (let node of nodes) {
-                    fragment.appendChild(node);
-                }
-
-                convertPastedContentFromWord(fragment);
-                this.editor.insertNode(fragment);
-                this.onPasteComplete(clipboardData);
-            });
+            restoreSnapshot(this.editor, clipboardData.snapshotBeforePaste);
         }
-    }
 
-    private onPasteComplete = (clipboardData: ClipBoardData) => {
-        let pasteHandler = this.pasteHandler || this.defaultPasteHandler;
-        if (clipboardData) {
-            // if any clipboard data exists, call into pasteHandler
-            pasteHandler(clipboardData);
+        switch (pasteOption) {
+            case PasteOption.PasteAsIs:
+                this.pasteHTML(clipboardData);
+                break;
+
+            case PasteOption.PasteText:
+                this.pasteText(clipboardData);
+                break;
+
+            case PasteOption.PasteWithCurrentFormat:
+                this.pasteWithCurrentFormat(clipboardData);
+                break;
         }
 
-        // add undo snapshot after paste
         // broadcast contentChangedEvent to ensure the snapshot actually gets added
         let contentChangedEvent: ContentChangedEvent = {
             eventType: PluginEventType.ContentChanged,
             source: 'Paste',
+            data: clipboardData,
         };
         this.editor.triggerEvent(contentChangedEvent, true /* broadcast */);
-        this.editor.addUndoSnapshot();
-    };
 
-    private defaultPasteHandler = (clipboardData: ClipBoardData) => {
-        let file = clipboardData.imageData ? clipboardData.imageData.file : null;
-        if (file) {
-            insertImage(this.editor, file);
-        }
+        // add undo snapshot after paste
+        this.editor.addUndoSnapshot();
     };
 
-    private retrieveHtml(editor: Editor, callback: (container: HTMLElement) => void) {
-        // cache original selection range in editor
-        let originalSelectionRange = editor.getSelectionRange();
-
-        if (!this.pasteContainer || !this.pasteContainer.parentNode) {
-            this.pasteContainer = fromHtml(CONTAINER_HTML, editor.getDocument())[0] as HTMLElement;
-            editor.insertNode(this.pasteContainer, {
-                position: ContentPosition.Outside,
-                updateCursor: false,
-                replaceSelection: false,
-                insertOnNewLine: false,
-            });
+    private pasteHTML(clipboardData: ClipBoardData, fragmentOverride?: DocumentFragment) {
+        if (clipboardData.textData || !clipboardData.imageData.file) {
```

Comment:
Why check textData instead of htmlData?

Reply:
Copy an image from OneNote, there HTML data will look like: <div><img></div>. In that case htmlData is not empty but textData is empty.

So if we only check htmlData, we will always paste in HTML instead of image.